Bug to fix: drwrap does not clean up all resources when exit

It looks like drwrap doesn't clean up all the resources when calling drwrap_exit() in client.
In might cause potential problems in clients when DR is actively invoked and stopped by the same user application several times in one run. Here is how I reproduce the problem in drcachesim client test case:
1. Call `drwrap_init()` inside `clients/drcachesim/tracer/tracer.cpp:drmemtrace_client_main`
2. Call `drwrap_exit()` inside `clients/drcachesim/tracer/tracer.pp:event_exit`
3. `include "drwrap.h"` in `clients/drcachesim/tracer/tracer.pp`, add `use_DynamoRIO_extension(${name} drwrap${ext_sfx})` under `clients/drcachesim/CMakeLists.txt:configure_DynamoRIO_client`
4. Build with test cases
5. `ctest -R .*burst_static` will fail 2 out of 3 times (test case: tool.drcacheoff.burst_static).

Sure! Here is the output of running `ctest --verbose -R \.*burst_static$` under debug build:
```
UpdateCTestConfiguration  from :/usr/local/google/home/yelouis/prj/Work/dynamorio/build_debug/DartConfiguration.tcl
Parse Config file:/usr/local/google/home/yelouis/prj/Work/dynamorio/build_debug/DartConfiguration.tcl
UpdateCTestConfiguration  from :/usr/local/google/home/yelouis/prj/Work/dynamorio/build_debug/DartConfiguration.tcl
Parse Config file:/usr/local/google/home/yelouis/prj/Work/dynamorio/build_debug/DartConfiguration.tcl
Test project /usr/local/google/home/yelouis/prj/Work/dynamorio/build_debug
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 244
    Start 244: code_api|tool.drcacheoff.burst_static

244: Test command: /usr/bin/cmake "-D" "precmd=foreach@/usr/bin/cmake@-E@remove_directory@drmemtrace.tool.drcacheoff.burst_static.*.dir" "-D" "cmd=/usr/local/google/home/yelouis/prj/Work/dynamorio/build_debug/clients/bin64/tool.drcacheoff.burst_static" "-D" "postcmd=firstglob@/usr/local/google/home/yelouis/prj/Work/dynamorio/build_debug/clients/bin64/drcachesim@-indir@drmemtrace.tool.drcacheoff.burst_static.*.dir" "-D" "postcmd2=" "-D" "postcmd3=" "-D" "cmp=/usr/local/google/home/yelouis/prj/Work/dynamorio/build_debug/suite/tests/offline-burst_static.expect" "-P" "/usr/local/google/home/yelouis/prj/Work/dynamorio/suite/tests/runmulti.cmake"
244: Test timeout computed to be: 1500
244: Running |/usr/bin/cmake;-E;remove_directory /usr/local/google/home/yelouis/prj/Work/dynamorio/build_debug/suite/tests/drmemtrace.tool.drcacheoff.burst_static.249608.1806.dir|
244: Running |/usr/bin/cmake;-E;remove_directory /usr/local/google/home/yelouis/prj/Work/dynamorio/build_debug/suite/tests/drmemtrace.tool.drcacheoff.burst_static.249608.2579.dir|
244: Running |/usr/bin/cmake;-E;remove_directory /usr/local/google/home/yelouis/prj/Work/dynamorio/build_debug/suite/tests/drmemtrace.tool.drcacheoff.burst_static.249608.3756.dir|
244: Running cmd |/usr/local/google/home/yelouis/prj/Work/dynamorio/build_debug/clients/bin64/tool.drcacheoff.burst_static|
244: Running postcmd |/usr/local/google/home/yelouis/prj/Work/dynamorio/build_debug/clients/bin64/drcachesim;-indir;/usr/local/google/home/yelouis/prj/Work/dynamorio/build_debug/suite/tests/drmemtrace.tool.drcacheoff.burst_static.249655.4415.dir|
244: CMake Error at /usr/local/google/home/yelouis/prj/Work/dynamorio/suite/tests/runmulti.cmake:130 (message):
244:   output |pre-DR init
244: 
244:   pre-DR start
244: 
244:   pre-DR detach
244: 
244:   DynamoRIO statistics:
244: 
244:                 Peak threads under DynamoRIO control :                 1
244:                                 Threads ever created :                 1
244:                                    System calls, pre :                 1
244:                                   System calls, post :                 1
244:                      Basic block fragments generated :               276
244:                            Trace fragments generated :                18
244:                Peak fcache combined capacity (bytes) :             94208
244:                       Peak fcache units on live list :                 4
244:                       Peak fcache units on free list :                 4
244:                   Peak special heap capacity (bytes) :             16384
244:                         Peak heap units on live list :                11
244:                         Peak heap units on free list :                 5
244:                            Peak client raw mmap size :             36864
244:                          Peak stack capacity (bytes) :            147456
244:                           Peak heap capacity (bytes) :            544768
244:                    Peak total memory from OS (bytes) :           1449984
244:                             Peak vmm blocks for heap :               213
244:                            Peak vmm blocks for cache :                64
244:                            Peak vmm blocks for stack :                42
244:                     Peak vmm blocks for special heap :                25
244:                     Peak vmm blocks for special mmap :                10
244:               Peak vmm virtual memory in use (bytes) :           1449984
244: 
244:   all done
244: 
244:   pre-DR init
244: 
244:   pre-DR start
244: 
244:   pre-DR detach
244: 
244:   DynamoRIO statistics:
244: 
244:                 Peak threads under DynamoRIO control :                 1
244:                                 Threads ever created :                 2
244:                                    System calls, pre :                 2
244:                                   System calls, post :                 2
244:                      Basic block fragments generated :               552
244:                            Trace fragments generated :                36
244:                Peak fcache combined capacity (bytes) :             94208
244:                       Peak fcache units on live list :                 4
244:                       Peak fcache units on free list :                 4
244:                   Peak special heap capacity (bytes) :             16384
244:                         Peak heap units on live list :                11
244:                         Peak heap units on free list :                 5
244:                            Peak client raw mmap size :             36864
244:                          Peak stack capacity (bytes) :            204800
244:                           Peak heap capacity (bytes) :            544768
244:                    Peak total memory from OS (bytes) :           1449984
244:                             Peak vmm blocks for heap :               213
244:                            Peak vmm blocks for cache :                64
244:                            Peak vmm blocks for stack :                58
244:                     Peak vmm blocks for special heap :                25
244:                     Peak vmm blocks for special mmap :                10
244:               Peak vmm virtual memory in use (bytes) :           1449984
244: 
244:   all done
244: 
244:   pre-DR init
244: 
244:   pre-DR start
244: 
244:   pre-DR detach
244: 
244:   DynamoRIO statistics:
244: 
244:                 Peak threads under DynamoRIO control :                 1
244:                                 Threads ever created :                 3
244:                                    System calls, pre :                 3
244:                                   System calls, post :                 3
244:                      Basic block fragments generated :               828
244:                            Trace fragments generated :                54
244:                Peak fcache combined capacity (bytes) :             94208
244:                       Peak fcache units on live list :                 4
244:                       Peak fcache units on free list :                 4
244:                   Peak special heap capacity (bytes) :             16384
244:                         Peak heap units on live list :                11
244:                         Peak heap units on free list :                 5
244:                            Peak client raw mmap size :             36864
244:                          Peak stack capacity (bytes) :            262144
244:                           Peak heap capacity (bytes) :            544768
244:                    Peak total memory from OS (bytes) :           1449984
244:                             Peak vmm blocks for heap :               213
244:                            Peak vmm blocks for cache :                64
244:                            Peak vmm blocks for stack :                74
244:                     Peak vmm blocks for special heap :                25
244:                     Peak vmm blocks for special mmap :                10
244:               Peak vmm virtual memory in use (bytes) :           1449984
244: 
244:   all done
244: 
244:   Cache simulation results:
244: 
244:   Core #0 (1 thread(s))
244: 
244:     L1I stats:
244:       Hits:                                0
244:       Misses:                              0
244:       Invalidations:                       0
244:     L1D stats:
244:       Hits:                                0
244:       Misses:                              0
244:       Invalidations:                       0
244: 
244:   Core #1 (0 thread(s))
244: 
244:   Core #2 (0 thread(s))
244: 
244:   Core #3 (0 thread(s))
244: 
244:   LL stats:
244: 
244:       Hits:                                0
244:       Misses:                              0
244:       Invalidations:                       0
244: 
244:   | failed to match expected output |pre-DR init
244: 
244:   pre-DR start
244: 
244:   pre-DR detach
244: 
244:   DynamoRIO statistics:
244: 
244:    *Peak threads under DynamoRIO control : *1
244: 
244:   .*
244: 
244:   all done
244: 
244:   pre-DR init
244: 
244:   pre-DR start
244: 
244:   pre-DR detach
244: 
244:   DynamoRIO statistics:
244: 
244:    *Peak threads under DynamoRIO control : *1
244: 
244:   .*
244: 
244:   all done
244: 
244:   pre-DR init
244: 
244:   pre-DR start
244: 
244:   pre-DR detach
244: 
244:   DynamoRIO statistics:
244: 
244:    *Peak threads under DynamoRIO control : *1
244: 
244:   .*
244: 
244:   all done
244: 
244:   Cache simulation results:
244: 
244:   Core #0 \(1 thread\(s\)\)
244: 
244:     L1I stats:
244:       Hits:                         *[0-9,\.]*.....
244:       Misses:                       *[0-9,\.]*...
244:       Invalidations:                *0
244: 
244:   .* Miss rate: 0[,\.]..%
244: 
244:     L1D stats:
244:       Hits:                         *[0-9,\.]*.....
244:       Misses:                       *[0-9,\.]*.
244:       Invalidations:                *0
244: 
244:   .* Miss rate: 0[,\.]..%
244: 
244:   Core #1 \(0 thread\(s\)\)
244: 
244:   Core #2 \(0 thread\(s\)\)
244: 
244:   Core #3 \(0 thread\(s\)\)
244: 
244:   LL stats:
244: 
244:       Hits:                         *[0-9,\.]*.
244:       Misses:                       *[0-9,\.]*..
244:       Invalidations:                *0
244: 
244:   .* Local miss rate: *[0-9]*[,\.]..%
244: 
244:       Child hits:                   *[0-9,\.]*...
244:       Total miss rate:                  [0-1][,\.]..%
244: 
244:   |
244: 
244: 
1/1 Test #244: code_api|tool.drcacheoff.burst_static ...***Failed    1.50 sec

0% tests passed, 1 tests failed out of 1

Total Test time (real) =   2.97 sec

The following tests FAILED:
	244 - code_api|tool.drcacheoff.burst_static (Failed)
Errors while running CTest
```

--- a/ext/drwrap/drwrap.c
+++ b/ext/drwrap/drwrap.c
@@ -970,7 +970,7 @@ void
 drwrap_exit(void)
 {
     /* handle multiple sets of init/exit calls */
-    int count = dr_atomic_add32_return_sum(&drwrap_init_count, -1);
+    int i, count = dr_atomic_add32_return_sum(&drwrap_init_count, -1);
     if (count != 0)
         return;
 

--- a/ext/drwrap/drwrap.c
+++ b/ext/drwrap/drwrap.c
@@ -970,7 +970,7 @@ void
 drwrap_exit(void)
 {
     /* handle multiple sets of init/exit calls */
-    int i, count = dr_atomic_add32_return_sum(&drwrap_init_count, -1);
+    int count = dr_atomic_add32_return_sum(&drwrap_init_count, -1);
     if (count != 0)
         return;
 
@@ -981,7 +981,7 @@ drwrap_exit(void)
         !dr_unregister_delete_event(drwrap_fragment_delete))
         ASSERT(false, "failed to unregister in drwrap_exit");
 
-    for (i = 0; i < POSTCALL_CACHE_SIZE; i++) {
+    for (int i = 0; i < POSTCALL_CACHE_SIZE; i++) {
         postcall_cache[i] = NULL;
     }
     postcall_cache_idx = 0;

--- a/ext/drwrap/drwrap.c
+++ b/ext/drwrap/drwrap.c
@@ -981,6 +981,11 @@ drwrap_exit(void)
         !dr_unregister_delete_event(drwrap_fragment_delete))
         ASSERT(false, "failed to unregister in drwrap_exit");
 
+    for (i = 0; i < POSTCALL_CACHE_SIZE; i++) {
+        postcall_cache[i] = NULL;
+    }
+    postcall_cache_idx = 0;
+
     hashtable_delete(&replace_table);
     hashtable_delete(&replace_native_table);
     hashtable_delete(&wrap_table);

